generator/src/main/scala/grpcgateway/generators/GatewayGenerator.scala

```diff
@@ -156,17 +157,21 @@ object GatewayGenerator extends protocbridge.ProtocCodeGenerator with Descriptor
       case PatternCase.POST =>
         printer
           .add(s"""case ("POST", "${http.getPost}") => """)
+          .add("for {")
           .addIndented(
-            s"val input = Try(JsonFormat.fromJsonString[${method.getInputType.getName}](body))",
-            s"Future.fromTry(input).flatMap(stub.$methodName)"
+            s"""msg <- Future(JsonFormat.fromJson[${method.getInputType.getName}](body)).recover { case _ => throw new Exception("Wrong json input. Check proto file") }""",
+            s"res <- stub.$methodName(msg)"
           )
+          .add("} yield res")
       case PatternCase.PUT =>
         printer
           .add(s"""case ("PUT", "${http.getPut}") => """)
+          .add("for {")
           .addIndented(
-            s"val input = Try(JsonFormat.fromJsonString[${method.getInputType.getName}](body))",
-            s"Future.fromTry(input).flatMap(stub.$methodName)"
+            s"""msg <- Future(JsonFormat.fromJson[${method.getInputType.getName}](body)).recover { case _ => throw new Exception("Wrong json input. Check proto file") }""",
```

Comment:
All these `throw new Exception(...)` generated an `INTERNAL_SERVER_ERROR` in the response, right? a `BAD_REQUEST` seems more appropriate in this case

Reply:
Yes, because throwable2HttpResponseStatus will return INERTNAL_SERVER_ERROR
I will fix it